### PR TITLE
Characteristic function bk

### PR DESCRIPTION
**Pull request by billy4all**
### What does this PR do?
This simple PR wants to add the option of returning the Characteristic Function of the Baer-Kradolfer picker implementation inside obspy (obspy.signal.trigger.pk_baer). I modified the C-source code and the python wrapper accordingly to obtain it.

### Why was it initiated?  Any relevant Issues?
No major issue, or problem. Just needed this ability for a project myself, I think it would be useful to the community as well.

### PR Checklist
- [x] Correct base branch selected? `master` for new features, `maintenance_...` for bug fixes
- [x] This PR is not directly related to an existing issue (which has no PR yet).
- [x] If the PR is making changes to documentation, docs pages can be built automatically.
      Just remove the space in the following string after the + sign: "+DOCS"
- [x] If any network modules should be tested for the PR, add them as a comma separated list
      (e.g. `clients.fdsn,clients.arclink`) after the colon in the following magic string: "+TESTS:"
      (you can also add "ALL" to just simply run all tests across all modules)
- [x] All tests still pass.
- [x] Any new features or fixed regressions are be covered via new tests.
- [x] Any new or changed features have are fully documented.
- [x] Significant changes have been added to `CHANGELOG.txt` .
- [x] First time contributors have added your name to `CONTRIBUTORS.txt` .

### CODE REVIEW

**Comment by d-chambers** [comment]
Hey @billy4all, 

This looks like a nice contribution, thanks for making it. 

It looks like circle CI doesn't like your naming of the characteristic function:

> obspy/signal/trigger.py:415:5: N806 variable 'CF' in function should be lowercase

So perhaps just rename it to `cf` or `char_func` and see if circle turns green.

File: CHANGELOG.txt [hunk 1]
```diff
@@ -67,6 +67,8 @@ master:
     * added read support for FOCMEC 'out' and 'lst' files (see #2156)
  - obspy.signal.trigger:
     * fix a bug in AR picker (see #2157)
+    * option to return Baer-Kradolfer characteristic function from pk_mbaer
+      function added
```
Comment (krischer) [inline CHANGELOG.txt]:
```suggestion
      function added (see #2341)
```

File: obspy/CONTRIBUTORS.txt [hunk 2]
```diff
@@ -92,3 +92,4 @@ Winkelman, Andrew
 Zaccarelli, Riccardo
 Zad, Seyed Kasra Hosseini
 Zhu, Lijun
+Bagagli, Matteo
```
Comment (krischer) [inline obspy/CONTRIBUTORS.txt]: Please adhere to the alphabetic ordering.

File: obspy/signal/src/pk_mbaer.c [hunk 3]
```diff
@@ -1,17 +1,20 @@
 /*--------------------------------------------------------------------
 # Filename: pk_mbaer.c
-#  Purpose: C-Verions of Baer Picker based on Fortran code of Baer
+#  Purpose: C-Version of Baer Picker based on Fortran code of Baer
 #   Author: Andreas Rietbrock, Joachim Wassermann
 # Copyright (C) A. Rietbrock, J. Wassermann
+#
+# Modified: M.Bagagli 02-2019
+            - The function now returns also the characteristic function
```
Comment (krischer) [inline obspy/signal/src/pk_mbaer.c]: This is not necessary. `git` keeps track of all changes an the reasoning behind it.

File: obspy/signal/src/pk_mbaer.c [hunk 4]
```diff
@@ -92,10 +95,11 @@ int tdownmax,tupevent;
 float thrshl1,thrshl2;
 int preset_len;
 int p_dur;
-{
+float *CF  //MB
```
Comment (krischer) [inline obspy/signal/src/pk_mbaer.c]: Unnecessary comment.

File: obspy/signal/src/pk_mbaer.c [hunk 5]
```diff
@@ -92,10 +95,11 @@ int tdownmax,tupevent;
...
+
 */
 
-int ppick (float *reltrc, int npts, int *pptime, char *pfm, float samplespersec, int tdownmax, int tupevent, float thrshl1, float thrshl2, int preset_len, int p_dur){
+int ppick (float *reltrc, int npts, int *pptime, char *pfm, float samplespersec, int tdownmax, int tupevent, float thrshl1, float thrshl2, int preset_len, int p_dur,float *CF){
```
Comment (krischer) [inline obspy/signal/src/pk_mbaer.c]:
```suggestion
int ppick (float *reltrc, int npts, int *pptime, char *pfm, float samplespersec, int tdownmax, int tupevent, float thrshl1, float thrshl2, int preset_len, int p_dur,float *cf){
```

File: obspy/signal/src/pk_mbaer.c [hunk 6]
```diff
@@ -117,6 +121,9 @@ int ppick (float *reltrc, int npts, int *pptime, char *pfm, float samplespersec,
       float min,max;
       float scale;
 
+      int CF_COUNTER=0;    //MB
```
Comment (krischer) [inline obspy/signal/src/pk_mbaer.c]:
```suggestion
      int cf_counter = 0;  /* MB */
```

File: obspy/signal/src/pk_mbaer.c [hunk 7]
```diff
@@ -265,10 +272,13 @@ int ppick (float *reltrc, int npts, int *pptime, char *pfm, float samplespersec,
...
       omega= y2/yt;
 
-      if(sdev > 0)
+      if(sdev > 0) {
           edev= (edat-mean)/sdev;  /* corresponds to CF(t), mean corresponds
                                       to S(t)
                                    */
+          CF[CF_COUNTER]=edev ;    //MB
+          CF_COUNTER++ ;           //MB
+      }                            //MB add brackets {}.
```
Comment (krischer) [inline obspy/signal/src/pk_mbaer.c]: Please lowercase everything and use `/* */` style comments to retain a consistent code style in the file.

File: obspy/signal/trigger.py [hunk 8]
```diff
@@ -394,9 +394,13 @@ def pk_baer(reltrc, samp_int, tdownmax, tupevent, thr1, thr2, preset_len,
         SF(t) on preset()
     :param p_dur: p_dur defines the time interval for which the maximum
         amplitude is evaluated Originally set to 6 secs
-    :return: (pptime, pfm) pptime sample number of parrival; pfm direction
-        of first motion (U or D)
-
+    :type return_cf: bool
+    :param return_cf: If ``True``, also return the charachteristic function
```
Comment (krischer) [inline obspy/signal/trigger.py]:
```suggestion
    :param return_cf: If ``True``, also return the characteristic function.
```

File: obspy/signal/trigger.py [hunk 9]
```diff
@@ -394,9 +394,13 @@ def pk_baer(reltrc, samp_int, tdownmax, tupevent, thr1, thr2, preset_len,
...
+        calculated by the C-routine.
```
Comment (krischer) [inline obspy/signal/trigger.py]: Remove this line. No need to mention implementation details in the docstring.

File: obspy/signal/trigger.py [hunk 10]
```diff
@@ -394,9 +394,13 @@ def pk_baer(reltrc, samp_int, tdownmax, tupevent, thr1, thr2, preset_len,
...
+    :return: (pptime, pfm [,cf]) pptime sample number of parrival;
+        pfm direction of first motion (U or D), optionally also the
+        numpy.ndarray float32 containing the values of the characteristic
```
Comment (krischer) [inline obspy/signal/trigger.py]:
```suggestion
        characteristic function.
```

File: obspy/signal/trigger.py [hunk 11]
```diff
@@ -406,16 +410,24 @@ def pk_baer(reltrc, samp_int, tdownmax, tupevent, thr1, thr2, preset_len,
     pfm = C.create_string_buffer(b"     ", 5)
     # be nice and adapt type if necessary
     reltrc = np.ascontiguousarray(reltrc, np.float32)
+    # Initiliaze CF array (MB)
+    c_float_p = C.POINTER(C.c_float)
+    cf_arr = np.ascontiguousarray(np.zeros(len(reltrc) - 1), np.float32)
```
Comment (krischer) [inline obspy/signal/trigger.py]:
```suggestion
    cf_arr = np.zeros(len(reltrc) - 1, dtype=np.float32, order="C")
```

This should be the same result but less internal copies and thus faster.

File: obspy/signal/src/pk_mbaer.c (hunk 6 above)
Comment (QuLogic) [reply obspy/signal/src/pk_mbaer.c]: Don't need the `MB`.

File: obspy/signal/src/pk_mbaer.c (hunk 7 above)
Comment (QuLogic) [reply obspy/signal/src/pk_mbaer.c]: Don't need the comment about brackets, nor the `MB` ones; `git blame` can provide that information.

**Comment by billy4all** [comment]
Ok, thanks a lot for the comments! Pushing the final modded branch

File: CHANGELOG.txt [hunk 12]
```diff
@@ -67,6 +67,7 @@ master:
     * added read support for FOCMEC 'out' and 'lst' files (see #2156)
  - obspy.signal.trigger:
     * fix a bug in AR picker (see #2157)
+    * function added (see #2341)
```
Comment (megies) [inline CHANGELOG.txt]: this got garbled up somehow, I think you deleted a line?

Comment (billy4all) [reply CHANGELOG.txt]: Yep I did, accidentally ... 
pushing the updated one now

**Comment by megies** [comment]
Thanks @billy4all and welcome on the contributors list! :tada:

**Pull request merged**